File: osm/astar.pl
#! /usr/bin/perl -w

use strict;
use OSM::Map;
use Data::Dumper;
use Storable;

my $dbfile = "localdatabase.data";

sub reconstruct_path {
    my $came_from = shift;
    my $goes_to = shift;
    my $current_node = shift;
    
    my @path = ();
    my @rev;
    my $cn = $current_node;
    
    while (defined($came_from->{$cn})) {
	$cn = $came_from->{$cn};
        push @rev,$cn;
    }
    while ($cn=pop(@rev)) {
	push @path,$cn;
    }
    push @path,$current_node;
    my $x = $current_node;
    if (defined($goes_to)) {
	while (exists($$goes_to{$x})) {
	    push @path,$$goes_to{$x};
	    $x=$$goes_to{$x};
	}
    }
    return  @path;
}

sub Astar {
    my $map = shift;
    my $startlat = shift;
    my $startlon = shift;
    my $goallat  = shift;
    my $goallon  = shift;
    my $vehicle = shift;
    
    my (%closedset,%goalset,%startset,%gs_score,%hs_score,%fs_score,%ds_score,%gg_score,%hg_score,%fg_score,%dg_score,%came_from,%goes_to);
    my (@openset,@f,@g,@h,@d,@to,@gl);
    
    $map->initRoute($vehicle);
    
    $map->fetchCoor($startlat,$startlon,1) unless $map->inboundCoor($startlat,$startlon);
    $map->fetchCoor($goallat,$goallon,1) unless $map->inboundCoor($goallat,$goallon);
    my $start = $map->findNode($startlat,$startlon,10);
    my $goal = $map->findNode($goallat,$goallon,10);
    print "start = $start, goal = $goal ".(defined($vehicle)?$vehicle:"")."\n";

    $startset{$start} = 1;
    $goalset{$goal} = 2;
    
    $gs_score{$start} = 0;
    $fs_score{$start} = $hs_score{$start} = $map->calc_h_score($start,$goal,$vehicle);
    $ds_score{$start} = $map->distance($start,$goal);
    
    $gg_score{$goal} = 0;
    $fg_score{$goal} = $hg_score{$goal} = $map->calc_h_score($goal,$start,$vehicle);
    $dg_score{$goal} = $map->distance($goal,$start);
    
    
    $openset[1] = \%startset;
    $openset[2] = \%goalset;
    $f[1] = \%fs_score;
    $f[2] = \%fg_score;
    $g[1] = \%gs_score;
    $g[2] = \%gg_score;
    $h[1] = \%hs_score;
    $h[2] = \%hg_score;
    $d[1] = \%ds_score;
    $d[2] = \%dg_score;
    $to[1] = \%came_from;
    $to[2] = \%goes_to;
    $gl[1] = $goal;
    $gl[2] = $start;
    
    while (keys(%startset) != 0 or keys(%goalset) != 0) {
	for (my $i=1;$i<=2;$i++) {
	    my $goal = $gl[$i];
	    my $xs = undef;
	    
	    for my $k (keys(%{$openset[$i]})) {
		$xs=$k unless defined($xs);
		$xs=$k if (${$f[$i]}{$k} < ${$f[$i]}{$xs});
	    }
	    if (defined($xs)) {
		if (exists($closedset{$xs}) and $closedset{$xs} != $i) {
		    my @p = reconstruct_path($to[1],$to[2],$xs);
		    return @p;
		}
		delete(${$openset[$i]}{$xs});
		$closedset{$xs} = $i;
#	        print "close $xs ".$f_score{$xs}."\n";                
		for my $y ($map->neighbours($xs)) {
		    $map->fetchNode($y) unless $map->inboundNode($y);
		    next if (defined($closedset{$y}));
                    my $prevnode = ${$to[$i]}{$xs};
		    my $tentative_g_score = ${$g[$i]}{$xs} + ($i==1?$map->cost($xs,$y,$prevnode):$map->cost($y,$xs,$prevnode));
		    my $tentative_is_better;
		    unless (defined(${$openset[$i]}{$y})) {
			${$openset[$i]}{$y} = 1 ;
			$tentative_is_better = 1;
		    } elsif ($tentative_g_score < ${$g[$i]}{$y}) {
			$tentative_is_better = 1;
		    } else {
			$tentative_is_better = 0;
		    }
		    if ($tentative_is_better) {
			${$to[$i]}{$y} = $xs;
			${$g[$i]}{$y} = $tentative_g_score;
			${$h[$i]}{$y} = $map->calc_h_score($y,$goal);
			${$d[$i]}{$y} = $map->distance($y,$goal);
			${$f[$i]}{$y} = ${$g[$i]}{$y}+${$h[$i]}{$y};
#			my $w = $map->getway($xs,$y);
#			print "$i $xs $y ",${$g[$i]}{$y}," ",${$h[$i]}{$y}," ",${$f[$i]}{$y},"\n" if $i==1;
#			print "$i $y $xs ",${$g[$i]}{$y}," ",${$h[$i]}{$y}," ",${$f[$i]}{$y},"\n" if $i==2;
#		        print "openset=",join(", ",keys(%openset)),"\n";
		    }
		}
	    }
	}
    }
    
#     return failure
    return "foutje";
}

sub print_path {
    my $map = shift;
    my @p1=@_;
    print join(", ",@p1);
    print " ",$map->node($p1[0])->{length} if defined($map->node($p1[0])->{length});
    print "\n";
    my $oldp;
    my $oldw;
    my $dir;
    my $olddir;
    my $totdist=0;
    for my $p (@p1) {
        my $w;

	print "$p -> ";
	my @ws = $map->getways($p);
	if ($#ws == 0) {
	    $w = $ws[0];
	} else {
	    if (defined($oldp)) {
	        $w=$map->getway($oldp,$p);
	    }
	}
	if (defined($w)) {
	    print $w,": ";
	    my $mwt = $map->way($w)->{tag};
	    print $$mwt{name}," " if defined($$mwt{name});
	    print $$mwt{highway}," " if defined($$mwt{highway});
	    print $$mwt{maxspeed}," " if defined($$mwt{maxspeed});
	    print $$mwt{ref}," " if defined($$mwt{ref});
	    my $d = $map->dist($oldp,$p);
	    if (defined($d)) {
                $olddir = $dir;
                $dir = $map->direction($oldp,$p);
		print "$dir $d ";
		$totdist += $d;
	    }
	    print $totdist," ";
            $oldw=$w;
            print $map->findLocation($p);
	}
	$oldp=$p;
	print "\n";
    }
}

my $arg = shift;
my $map;
if (defined($arg)) {
    $map = OSM::Map->new();
    if ($arg eq "local") {
        $map->useLocaldata("map.osm");
    }
} else {
    $map = retrieve $dbfile;
    $map->initialize();
}

###huis school
#print_path($map,Astar($map,52.297277,4.862030,52.29334,4.85876));
#print_path($map,Astar($map,52.2973969,4.8620826,52.2933,4.8588,'foot'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.2933,4.8588,'bicycle'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.2933,4.8588,'car'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.2935821,4.8593675,'car'));
#print_path($map,Astar($map,52.297275,4.8616077,52.2933,4.8588));
#print_path($map,Astar($map,52.297275,4.8616077,52.2933,4.8588,'foot'));
#print_path($map,Astar($map,52.297275,4.8616077,52.2933,4.8588,'bicycle'));
#print_path($map,Astar($map,52.297275,4.8616077,52.2933,4.8588,'car'));

###Huis hockey
#print_path($map,Astar($map,52.297275,4.8616077,52.2886,4.8508));
#print_path($map,Astar($map,52.297275,4.8616077,52.2886,4.8508,'foot'));
#print_path($map,Astar($map,52.297275,4.8616077,52.2886,4.8508,'bicycle'));
#print_path($map,Astar($map,52.297275,4.8616077,52.2886,4.8508,'car'));


#print_path($map,Astar($map,52.297275,4.8616077,52.311311,4.8468799,'car'));
#print_path($map,Astar($map,52.297275,4.8616077,52.4176385,4.8708703,'bicycle'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.278473,4.846854));
#print_path($map,Astar($map,52.2973969,4.8620826,52.278473,4.846854,'foot'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.278473,4.846854,'bicycle'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.278473,4.846854,'car'));
#print_path($map,Astar($map,52.297275,4.8616077,52.4176385,4.8708703,'car'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.2933,4.8588));
#print_path($map,Astar($map,52.297275,4.8616077,52.2935821,4.8593675));
#print_path($map,Astar($map,52.297275,4.8616077,52.2935821,4.8593675,'foot'));
#print_path($map,Astar($map,52.297275,4.8616077,52.2935821,4.8593675,'bicycle'));
#print_path($map,Astar($map,52.297275,4.8616077,52.2935821,4.8593675,'car'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.2932895,4.8544163));
#print_path($map,Astar($map,52.297275,4.8616077,52.2932895,4.8544163));
#print_path($map,Astar($map,52.297275,4.8616077,52.2871944,4.8503294,'bicycle'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.2871944,4.8503294,'car'));

###Coentunnel
#print_path($map,Astar($map,52.2973969,4.8620826,52.4184,4.8724));
#print_path($map,Astar($map,52.2973969,4.8620826,52.4184,4.8724,'foot'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.4184,4.8724,'car'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.4184,4.8724,'bicycle'));
#print_path($map,Astar($map,52.4184,4.8724,52.2973969,4.8620826,'car'));
#print_path($map,Astar($map,52.4184,4.8724,52.2973969,4.8620826,'bicycle'));

## Croon Delft
#print_path($map,Astar($map,52.2973969,4.8620826,51.9972199,4.3855367,'car'));
#print_path($map,Astar($map,52.2973969,4.8620826,51.9972199,4.3855367,'bicycle'));
#print_path($map,Astar($map,52.2973969,4.8620826,51.9972199,4.3855367));
#print_path($map,Astar($map,52.4184,4.8724,51.9972199,4.3855367,'car'));

#I# CT Barendrecht
#print_path($map,Astar($map,52.2973969,4.8620826,51.8503978,4.5091717,'car'));
#print_path($map,Astar($map,51.8503978,4.5091717,52.2973969,4.8620826,'car'));
#print_path($map,Astar($map,52.2973969,4.8620826,51.8503978,4.5091717,'bicycle'));

## ProRail
print_path($map,Astar($map,52.2973969,4.8620826,52.087473,5.115715,'car'));
print_path($map,Astar($map,52.2973969,4.8620826,52.087473,5.115715,'bicycle'));
print_path($map,Astar($map,52.087473,5.115715,52.2973969,4.8620826,'car'));

## Roquebrune
#print_path($map,Astar($map,52.2973969,4.8620826,43.4046930,6.6792379,'car'));

###Brussel
#print_path($map,Astar($map,52.2973969,4.8620826,50.8417207,4.3832422,'car'));

$map->saveOSMdata($dbfile);
store $map,$dbfile;

File: osm/astar.cpp
#include "osm.h"
#include <map>

namespace osm {
  using namespace std;
  
  void Map::initRoute(const string &vehicle) {
    if (vehicle != "" && _profiles.find(vehicle) == _profiles.end())
        throw domain_error("Onbekend voertuig : "+vehicle);
    _vehicle = vehicle;
  }

  double Map::cost(const long n1,const long n2,const long prevnode) { 
    double dist = distance(n1,n2);
    if (_vehicle == "") return dist;
    return dist;
  }

  double Map::calc_h_score(const long n1,const long n2) {
    double dist = distance(n1,n2);
    if (_vehicle == "") return dist;
    return dist;
  }
  
  long Map::AstarHelper(int set,long goal,set_type &openset,set_type &closedset,score_type &f,score_type &g,score_type &h,score_type &d,route_type &to) {
    long xs = 0;
    set_type::iterator k;

    for (k=openset.begin();k!=openset.end();k++) {
      if (xs == 0) 
	xs = k->first;
      if (f[k->first] < f[xs])
	xs = k->first;
    }
    if (xs == 0)  // no more nodes to process should not happen in our situation 
      throw runtime_error("xs = 0");

    k = closedset.find(xs);
    if (k != closedset.end() && k->second != set) {
      cout << "set = " << set << " xs match = " << xs << endl;
      return xs;
    }
    openset.erase(xs);
    closedset[xs]=set;
    vector<long> neighbours;
    vector<long>::iterator y;

    getNeighbours(xs,neighbours);
    for (y=neighbours.begin();y != neighbours.end();y++) {
      long xs1=xs;
      long prevnode = 0;
      route_type::iterator pn;
      pn = to.find(xs1);
      if (pn != to.end())
	prevnode = pn->second;
      k = closedset.find(*y);
      if (k == closedset.end()) {
	double tentative_g_score = g[xs1] + (set==1?cost(xs1,*y,prevnode):cost(*y,xs1,prevnode));
	bool tentative_is_better = false;
	k = openset.find(*y);
	if (k == openset.end()) {
	  openset[*y] = 1;
	  tentative_is_better = true;
	} else if (tentative_g_score < g[*y]) 
	  tentative_is_better = true;
	if (tentative_is_better) {
	  to[*y] = xs1;
	  g[*y] = tentative_g_score;
	  h[*y] = calc_h_score(*y,goal);
	  d[*y] = distance(*y,goal);
	  f[*y] = g[*y] + h[*y];
	  if (set == 1) 
	    cout << set << " " << xs << " " << *y << " " << g[*y] << " " << h[*y] << " " << f[*y] << endl;
	  else
	    cout << set << " " << *y << " " << xs << " " << g[*y] << " " << h[*y] << " " << f[*y] << endl;
	}
      }
    }
    return 0;
  }
  
  double Map::Astar(const long n1,const long n2,const string &vehicle,list<long> &route) {
    set_type startset,goalset,closedset;
    score_type gs_score,hs_score,fs_score,ds_score, gg_score,hg_score,fg_score,dg_score;
    route_type came_from,goes_to;
    long xs = 0;

    initRoute(vehicle);
    startset[n1] = 1;
    goalset[n2] = 1;
    gs_score[n1] = 0;
    dg_score[n2] = ds_score[n1] = distance(n1,n2);
    while (xs == 0 && (!startset.empty() || !goalset.empty())) {
      xs = AstarHelper(1,n2,startset,closedset,fs_score,gs_score,hs_score,ds_score,came_from);
      if (xs == 0)
        xs = AstarHelper(2,n1,goalset,closedset,fg_score,gg_score,hg_score,dg_score,goes_to);
    }

    route.clear();
    if (xs == 0)
        throw range_error("Niet mogelijk om een royute te berekenen");
    route.push_back(xs);
    route_type::iterator rp;

    long x = xs;
    while ((rp = came_from.find(x)) != came_from.end()) {
      x = rp->second;
      route.push_front(x);
    }
    x = xs;
    while ((rp = goes_to.find(x)) != goes_to.end()) {
      x = rp->second;
      route.push_back(x);
    }

    return 0;
  }
  
  double Map::Astar(const long n1,const double lat2,const double lon2,const string &vehicle,list<long> &route) {
    return 0;
  }
  
  double Map::Astar(const double lat1,const double lon1,const long n2,const string &vehicle,list<long> &route) {
    return 0;
  }
  
  double Map::Astar(const double lat1,const double lon1,const double lat2,const double lon2,const string &vehicle,list<long> &route) {
    return 0;
  }
}
